### PR TITLE
Fix #6395

### LINKED ISSUE #6395: Improve regex for semver in automation rules
We have a very basic pattern, but on the semver faq they have a better one

https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string

This needs to be changed and with a comment to the link from where it came from.

https://github.com/readthedocs/readthedocs.org/blob/ff0ad67991e599f79e9d956a3733785ccb78c505/readthedocs/builds/constants.py#L113-L113

### PR DESCRIPTION
**Pull request by piyushpalawat99**
This is my first pull request in this project. Please let me know if there's something that I need to fix.
PS: Created a new branch for ease of use.

### CODE REVIEW

**Comment by piyushpalawat99** [comment]
@stsewd I'm sorry to bother you but can you help me with this pr (it has not passed the Travis CI build)?

File: readthedocs/builds/constants.py [hunk 1]
```diff
@@ -110,7 +110,8 @@
 ALL_VERSIONS = 'all-versions'
 ALL_VERSIONS_REGEX = r'.*'
 SEMVER_VERSIONS = 'semver-versions'
-SEMVER_VERSIONS_REGEX = r'^v?(\d+\.)(\d+\.)(\d+)(-.+)?$'
+SEMVER_VERSIONS_REGEX = r'^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'
+# pattern referred from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
```
Comment (stsewd) [inline readthedocs/builds/constants.py]: You should put the comment above, not below.

File: readthedocs/builds/constants.py [hunk 2]
```diff
@@ -110,7 +110,8 @@
 ALL_VERSIONS = 'all-versions'
 ALL_VERSIONS_REGEX = r'.*'
 SEMVER_VERSIONS = 'semver-versions'
-SEMVER_VERSIONS_REGEX = r'^v?(\d+\.)(\d+\.)(\d+)(-.+)?$'
+SEMVER_VERSIONS_REGEX = r'^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'
```
Comment (stsewd) [inline readthedocs/builds/constants.py]:
```suggestion
SEMVER_VERSIONS_REGEX = r'^v?(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'
```

v should be optional

File: readthedocs/builds/constants.py [hunk 3]
```diff
@@ -110,7 +110,9 @@
 ALL_VERSIONS = 'all-versions'
 ALL_VERSIONS_REGEX = r'.*'
 SEMVER_VERSIONS = 'semver-versions'
-SEMVER_VERSIONS_REGEX = r'^v?(\d+\.)(\d+\.)(\d+)(-.+)?$'
+
+# Pattern referred from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+SEMVER_VERSIONS_REGEX = r'^v?(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'	# noqa
```
Comment (stsewd) [inline readthedocs/builds/constants.py]:
```suggestion
SEMVER_VERSIONS_REGEX = r'^v?(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'  # noqa

Use two spaces before the comment, not tabs
```

**Comment by piyushpalawat99** [comment]
Thanks for guiding me through it.

File: readthedocs/builds/constants.py [hunk 4]
```diff
@@ -110,7 +110,9 @@
 ALL_VERSIONS = 'all-versions'
 ALL_VERSIONS_REGEX = r'.*'
 SEMVER_VERSIONS = 'semver-versions'
-SEMVER_VERSIONS_REGEX = r'^v?(\d+\.)(\d+\.)(\d+)(-.+)?$'
+
+# Pattern referred from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
```
Comment (stsewd) [inline readthedocs/builds/constants.py]:
```suggestion
# Pattern referred from
# https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
```

This should fix the linter, the line was too long, but it makes sense for urls

**Comment by piyushpalawat99** [comment]
The issue still seems to be existing. Can this be a problem on Travis' side?

**Comment by stsewd** [comment]
@piyushpalawat99 you can check the travis log to find out the problem

> builds/constants.py
>  Line: 114
>    pep8: W291 / trailing whitespace (col 24)

Or run the tests locally https://docs.readthedocs.io/en/stable/development/tests.html

**Comment by piyushpalawat99** [comment]
@stsewd Thanks. Done.

**Comment by piyushpalawat99** [comment]
Is there something else that I should do to get it merged?

**Comment by piyushpalawat99** [comment]
@stsewd can you help me with closing this issue?

**Pull request merged**